### ISSUE TITLE
Tracking: Known issues with SODA

### ISSUE THREAD

**Issue by morgiyan** [oracle/node-oracledb#1088]
This is to track known [SODA](https://docs.oracle.com/en/database/oracle/simple-oracle-document-access/index.html) issues.

Client-libraries required for SODA production use: 18.5 and above.
Oracle Database: 18 and above.

**Known issues affecting SODA node.js are:**

- (1) **Bug 28533043** - this is an issue at OCI level affecting performance of "finds" that return multiple documents. Due to lack of reliance on OCI array fetch, which allows fetching multiple documents in one round-trip, fetching each document requires a round-trip to the database.  **Status**: fixed, now getting backported to 18.x and 19.x DBRU.

- (2) ~~**Bug 29607136** - this is an RDBMS-side bug is in index specification processing: when multiple varchar2 fields are indexed, an error is produced (instead of a successful index create). The error is in automatic computation of maxLength field of the index spec. There's an easy workaround for this bug, which is to specify the maxLength field explicitly for each varchar2 field of the index spec.~~  **Status**: fixed in DB 18.8 and 19.5.
 
- (3) ~~sodaCollection.metaData should be an Object (as documented) but is a String.~~ **Status**: fixed in node-oracledb 4.0.

- (4) ~~getOne() will fail if Oracle Client is < 19.5. This broke in node-oracledb 5.0.0.  The workaround (if you can't upgrade) is to explicitly set the fetch arraysize to 0: e.g. `doc = await collection.find().fetchArraySize(0).key(key).getOne();`~~  **Status**: Fixed in node-oracledb 5.1.

- (5) if you are using Oracle Database 21 (or later) and you create new collections, then check out the notes in [Node-oracledb SODA Requirements](https://oracle.github.io/node-oracledb/doc/api.html#sodarequirements).

**Enhancement Requests**

- Case insensitive sorting, see https://github.com/oracle/node-oracledb/issues/1099
- Need more examples.

**Comment by cjbj** [comment]
@morgiyan this is the place.

**Comment by cjbj** [comment]
I'll start: one issue is that `sodaCollection.metaData` should be an Object (as documented) but is a String.  This will change to an Object in node-oracledb 4.0

**Comment by charleaux** [comment]
Any idea when SODA will be ready for use in production?

**Comment by morgiyan** [comment]
Let me check with @cjbj, but I believe we should be removing the "preview" status any day now (it's just a matter of when Chris will be able to update the doc).

Known issues are:

**Bug 28533043** - this is a bug at OCI level affecting performance of "finds" that return multiple documents. Fixed, now getting backported to 18.x and 19.x series.

**Bug 29607136** - this is a bug in index specification, when multiple varchar2 fields are indexed, an error is produced (instead of a successful index create). The error is in automatic computation of maxLength field of the index spec. **There's an easy workaround for this bug,** which is to specify the maxLength field explicitly for each varchar2 field of the index spec. Bug is in the process of getting fixed, and will be backported to 18.x and 19.x series shortly.

**Comment by charleaux** [comment]
Awesome 👍 I can't wait

**Comment by morgiyan** [comment]
Charles (@charleaux), the preview status has been removed. Please see the list of known issues above (in particular issue 1, the other two are minor). If you need the fix for issue 1 to go production, let me know. Thanks!

**Comment by charleaux** [comment]
Is there any reason SODA isn't supported with node prior to OracleDB 18.3?

I keep seeing this in the examples:
 *   Requires Oracle Database and Client 18.3, or higher.

**Comment by cjbj** [comment]
@charleaux there are some bugs in the OCI layer not fixed until the 18.5 & 19.3 clients.  In particular a big memory leak. While SODA APIs in node-oracledb work with 18.3 client, you will hit these bugs.  Hence we want you to upgrade before going production.  Note these bugs do not impact the SODA Java API.

**Comment by morgiyan** [comment]
@cjbj, I think @charleaux might be asking about the database itself (as opposed to instant client).  @charleaux, is that right? If that's what's being asked, I can reply to that.

**Comment by charleaux** [comment]
Thank you for clarifying, you're right I was more referring to the database server side. I was more curious than anything.

**Comment by charleaux** [comment]
Thanks for making this happen!

I was able to shrink my code in many of my routes by ~40% in some cases switching to the soda api. 

Maybe at some point we can get more examples around SODA? I'd like to see more in the different kinds of indexing as well as pagination. It's not critical for me since I've been able to dig into the other non-node SODA documents to figure out how things work but the examples have been useful for me when learning and it's very likely there are others who would benefit from it.

**Comment by cjbj** [comment]
@charleaux PR's for new examples are welcome

**Comment by morgiyan** [comment]
Great to hear that @charleaux!

As @cjbj say PRs are welcome, and I'll also make a note for us to add examples as soon we have some free cycles.

**Comment by charleaux** [comment]
@morgiyan @cjbj 
Thanks guys, I know you're probably super busy as it is already. When things slow down on my end I'll see if if you still need any PRs for this also and come up with any I can think of.

**Comment by charleaux** [comment]
Just FYI in case anyone else was having these issues.

I was troubleshooting an issue where node would crash randomly and when it did crash sometimes this would show up in the logs (sometimes nothing would show up):
[Error: ORA-40736: Query-By-Example (QBE) filter is not valid JSON
JZN-00049: Input to JSON parser is empty] errorNum: 40736, offset: 0 }
[31m[nodemon] app crashed - waiting for file changes before starting...[39m

The filter wasn't anything complicated:
`let filter = {
    $and: [
        { measureId: { $eq: `${performanceDataItem.measureId}` } },
        { period: { $eq: `${performanceDataItem.period}` } }
    ]
};`

What resolved the issue for me was upgrading the instant client from 18_5 to 19_3.

**Comment by cjbj** [comment]
Tagging @morgiyan for comment on the ORA-40736.

**Comment by morgiyan** [comment]
@charleaux the error means that SODA thought the JSON that was passed in for the QBE is invalid.

Could you post a small complete reproducer so that we can check that the syntax of the QBE is correct?

The above, for example, seems to be missing backticks for the template literals, and I can't tell what values the template literals evaluate to.

Will look into it once you post a reproducer.

**Comment by cjbj** [comment]
@morgiyan I edited @charleaux's post and used triple backticks around the whole block; you can see the embedded ticks now.

**Comment by morgiyan** [comment]
Thanks @cjbj, still would like to get the values template literals would evaluate to.

**Comment by charleaux** [comment]
@morgiyan 
The measureID was just something like this: an ID field "00476D3241D64F6EBF203AEBDE285F00"
and the period i was using something simple to rule out special characters "Jan00"

**Comment by morgiyan** [comment]
Thanks @charleaux, we'll check (since we are not aware of any such issues with 18.5).

**Comment by morgiyan** [comment]
@charleaux we ran this exact QBE with 18.5 IC, and cannot reproduce the issue. Runs fine for us. Maybe you could double check that you're trying this exact QBE with 18.5 IC. Thanks.

**Comment by cjbj** [comment]
Oracle DB 20c is now available as a [preview](https://blogs.oracle.com/database/ga-of-oracle-database-20c-preview-release).  

By default in the preview, Oracle Database 20c SODA collections use a new [JSON storage type](https://docs.oracle.com/en/database/oracle/oracle-database/20/newft/new-json-data-type.html).   Since older Oracle Client libraries do not understand the new type, if you are using an older Oracle client library to access Oracle Database 20c, then use a work around:
- Set the database's [COMPATIBLE initialization parameter](https://www.oracle.com/pls/topic/lookup?ctx=dblatest&id=GUID-A2E90F08-BC9F-4688-A9D0-4A948DD3F7A9) to < 20 so the default type remains unchanged.
- Alternatively, if you want to set COMPATIBLE to 20 in order to take advantage of 20c features, then you will need to do one of the following:
      - never create any new collections
      - explicitly use metadata and choose the storage type when creating collections

There is no problem if you are using Oracle Client libraries 20.

**Comment by cjbj** [comment]
`getOne()` etc will fail if Oracle Client is < 19.5. This broke in node-oracledb 5.0.0.  The workaround (i.e for macOS users stuck with 19.3 client libraries) is to explicitly set the fetch arraysize to 0: e.g. `doc = await collection.find().fetchArraySize(0).key(key).getOne();`  We'll fix this in a future node-oracledb release.

**Comment by cjbj** [comment]
I've updated the original post with a link to info for users moving to Oracle DB 21.

**Comment by cjbj** [comment]
Playing with the SODA metadata cache feature of [node-oracledb 5.2.0-dev](https://github.com/oracle/node-oracledb/blob/master/CHANGELOG.md) shows a huge reduction in round-trips when repeatedly opening collections.  This equates to a reduced load on the DB and faster, more scalable applications.  The cache also needs Oracle Client 19.11 (or later) or Oracle Client 21.3+ (which isn't available yet), so to test it out, grab Oracle Client 19.11 and set `sodaMetaDataCache` to `true` when opening a connection pool.

Node-oracledb 5.2.0-dev also avoids a round-trip needed if a pooled connection has to check the DB version - which is done when using SODA (or AQ).  This optimization is for Oracle Client 19 and earlier.  Oracle Client 21.1 libraries already have an equivalent optimization.

Bottom line: if you're a SODA user, grab Instant Client 19.11 (even if you have 21.1) and try the SODA metadata cache in [node-oracledb 5.2.0-dev](https://oracle.github.io/node-oracledb/INSTALL.html#github) installed from the master branch.

**Comment by sharadraju** [comment]
Closing this as an issue
